service/csv/csv.go

```diff
@@ -0,0 +1,116 @@
+package csv
+
+import (
+	"bootcamp/domain/model"
+	"encoding/csv"
+	"errors"
+	"fmt"
+	"os"
+	"strconv"
+)
+
+type csvServiceInterface interface {
+	readCSV() (model.PokemonList, error)
+	writeCSV(pokemon model.FetchedPokemon) error
+}
+
+type CsvService struct{}
+
+func (s CsvService) readCSV() (model.PokemonList, error) {
+	var pokemonList model.PokemonList
+	recordFile, err := os.Open("assets/pokemon.csv")
+
+	if err != nil {
+		return nil, errors.New("Could not open file")
+	}
+
+	reader := csv.NewReader(recordFile)
+	records, err := reader.ReadAll()
+
+	if err != nil {
+		return nil, errors.New("Could not read file")
+	}
+
+	for _, pokemon := range records {
+		id, err := strconv.Atoi(pokemon[0])
+
+		if err != nil {
+			return pokemonList, errors.New("Cannot get id from row")
+		}
+
+		pkNumber, err := strconv.Atoi(pokemon[1])
+
+		if err != nil {
+			return pokemonList, errors.New("Cannot get pokedex number from row")
+		}
+
+		pk := model.Pokemon{
+			Id:            id,
+			PokedexNumber: pkNumber,
+			Name:          pokemon[2],
+			Types:         pokemon[3],
+			Region:        pokemon[4]}
+
+		pokemonList = append(pokemonList, pk)
+	}
+
+	err = recordFile.Close()
+
+	if err != nil {
+		return pokemonList, errors.New("Error while closing file")
+	}
+
+	return pokemonList, nil
+}
+
+func (s CsvService) writeCSV(pokemon model.FetchedPokemon) error {
+	file, err := os.Create("assets/fetched_pokemon.csv")
+
+	if err != nil {
+		return errors.New("Could not create file")
+	}
+
+	defer file.Close()
+	writer := csv.NewWriter(file)
+	defer writer.Flush()
+
+	pokemonId := strconv.Itoa(pokemon.Id)
+	pokemonHeight := strconv.Itoa(pokemon.Height)
+	pokemonOrder := strconv.Itoa(pokemon.Order)
+	pokemonWeight := strconv.Itoa(pokemon.Weight)
+
+	var pokemonForms string
+	for key, value := range pokemon.Forms {
+		pokemonForms = fmt.Sprintf("%v=\"%v\"", key, value)
+	}
+
+	pokemonString := []string{pokemonId, pokemonForms, pokemonHeight, pokemon.Name, pokemonOrder, pokemonWeight}
+
+	err = writer.Write(pokemonString)
+
+	if err != nil {
+		return errors.New("Could not write fetched pokemon")
+	}
+
+	return nil
+}
+
+var s csvServiceInterface
+
+func init() {
+	s = CsvService{}
+}
+
+/*
+ReadCSV reads a CSV with Pokemon information and transform the content to the PokemonList struct type
+*/
+func ReadCSV() (model.PokemonList, error) {
```

Comment:
This is the same as what I mentioned in the controller, it just seems like duplicating code.

Reply:
👌